src/instamatic/experiments/fast_adt/experiment.py

```diff
@@ -0,0 +1,452 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass, field
+from pathlib import Path
+from queue import Queue
+from threading import Thread
+from typing import Any, Dict, Iterator, List, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from matplotlib import pyplot as plt
+from typing_extensions import Self
+
+from instamatic import config
+from instamatic._collections import NullLogger
+from instamatic._typing import AnyPath
+from instamatic.calibrate import CalibBeamShift, CalibMovieDelays, CalibStageRotation
+from instamatic.calibrate.filenames import CALIB_BEAMSHIFT
+from instamatic.experiments.experiment_base import ExperimentBase
+from instamatic.processing.ImgConversionTPX import ImgConversionTPX as ImgConversion
+
+
+def safe_range(start: float, stop: float, step: float) -> np.ndarray:
+    """Find 2+ floats between `start` and `stop` (inclusive) ~`step` apart."""
+    step_count = max(round(abs(stop - start) / step) + 1, 2)
+    return np.linspace(start, stop, step_count, endpoint=True, dtype=float)
+
+
+class FastADTEarlyTermination(RuntimeError):
+    """Raised if FastADT experiment terminates early due to known reasons."""
+
+
+class FastADTMissingCalibError(RuntimeError):
+    """Raised if some calibration is strictly required but missing."""
+
+
+@dataclass
+class Step:
+    """A dataclass representing a single step in the experimental `Run`."""
+
+    Index: int
+    alpha: float
+    beamshift_x: Optional[float] = None
+    beamshift_y: Optional[float] = None
+    delta_x: Optional[float] = None
+    delta_y: Optional[float] = None
+    image: Optional[np.ndarray] = None
+    meta: dict = field(default_factory=dict)
+
+
+class Run:
+    """Collection of details of a generalized single FastADT run.
+
+    Includes a `.table` describing details of individual steps (to be) measured:
+      - alpha - average value of the rotation axes for given frame
+      - delta_x - x beam shift relative from center needed to track the crystal
+      - delta_y - y beam shift relative from center needed to track the crystal
+
+    Other class attributes include:
+      - continuous - whether the run involves continuous scan or single frames
+      - exposure - time spent collecting each frame, expressed in seconds
```

Comment:
Convert to numpydoc for consistency with the rest of the code:

```suggestion
	Attributes
	----------
	exposure : float
        Time spent collecting each frame, expressed in seconds
    continuous : bool
        Whether the run involves continuous scan (True) or single frames (False)
    table: pd.DataFrame
	    Describes details of individual steps (to be) measured:
	      - alpha - average value of the rotation axes for given frame
	      - delta_x - x beam shift relative from center needed to track the crystal
	      - delta_y - y beam shift relative from center needed to track the crystal
```